src/main.rs

```diff
@@ -38,13 +54,55 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     let addr = "0.0.0.0:4243".parse().unwrap();
     tracing::info!("chariott listening on {addr}");
 
-    let router = Server::builder()
-        .add_service(ChariottServiceServer::new(ChariottServer::new(registry, broker)));
+    let server = Arc::new(ChariottServer::new(registry, broker));
+    let router =
+        Server::builder().add_service(ChariottServiceServer::from_arc(Arc::clone(&server)));
 
     #[cfg(build = "debug")]
     let router = router.add_service(reflection_service);
 
-    router.serve_with_ctrl_c_shutdown(addr).await?;
+    let error_cancellation_token = CancellationToken::new();
+    let ctrl_c_cancellation_token = ctrl_c_cancellation();
+
+    let prune_loop = {
+        use tokio::{select, spawn, time::sleep};
+
+        let ctrl_c_cancellation_token = ctrl_c_cancellation_token.clone();
```

Comment:
[nit] `child_token` would make it more explicit that this is for consuming cancellations and not for generating them, as the tree from the child token cannot affect the parent.

```suggestion
        let ctrl_c_cancellation_token = ctrl_c_cancellation_token.child_token();
```

Reply:
The side effect of the prune loop accidentally cancelling the `ctrl_c_cancellation_token` would also shut down the server, which is acceptable in this scenario. Consider this comment resolved.